### ISSUE TITLE
uwsgi: Target is Python 0, which is not supported by this pyflame build.

### ISSUE THREAD

**Issue by zarnovican** [uber/pyflame#45]
Hi,

I'm getting "Target is Python 0, which is not supported by this pyflame build." error when running `pyflame` against `uwsgi` (despite what the issue #23 says, that it should be working).

OS: Ubuntu Xenial 16.04
Python: 2.7
uwsgi: 2.0.14
pyflame: 1.2.1

Steps to reproduce:
* the same as #23 

Running `uwsgi`:
```bash
$ uwsgi --workers 1 --module app:app --http-socket=127.0.0.1:8011
...
Python version: 2.7.12 (default, Nov 19 2016, 06:48:10)  [GCC 5.4.0 20160609]
...
```

Running `pyflame`:
```bash
$ sudo ./pyflame/src/pyflame `pgrep uwsgi`
Target is Python 0, which is not supported by this pyflame build.
$
```

uwsgi really is loading libpython2.7
```bash
$ sudo lsof -p `pgrep uwsgi` | grep libpython
uwsgi   11149 ubuntu  mem       REG  253,1  3582904  19947 /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
$
```

Looking at the `src/pyfrob.cc`, you seem to fallback to "libpython2.7.so", which does exist on my system:
```bash
$ ldconfig -p | grep libpython
	libpython3.5m.so.1.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpython3.5m.so.1.0
	libpython2.7.so.1.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0
	libpython2.7.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libpython2.7.so
$
```

I have also tried uwsgi options `--enable-threads`, `--single-interpreter`, `--master/workers` with no luck.

**Comment by zarnovican** [comment]
In `ELF::GetThreadState(PyVersion *version)` (https://github.com/uber/pyflame/blob/master/src/symbol.cc#L116) it is assumed that in ELF symbol table, there is `_PyThreadState_Current` and one of `PyString_Type` (in Py2) or `PyBytes_Type` (in Py3).

In my case, `uwsgi` has only the TreadState
```
$ objdump -t /proc/`pgrep uwsgi`/exe | grep _PyThreadState_Current
0000000000737760 g     O .bss	0000000000000008              _PyThreadState_Current
$ objdump -t /proc/`pgrep uwsgi`/exe | grep PyString_Type
$ objdump -t /proc/`pgrep uwsgi`/exe | grep PyBytes_Type
$
```
Therefor the function will return the threadstate address without setting the version (remains "0").

In `ThreadStateAddr` (https://github.com/uber/pyflame/blob/master/src/pyfrob.cc#L78) that will cause to return early and the fallback code to read "libpython2.7.so" is never reached.

I have tried to fix it by exiting from `GetThreadState` only when _both_ threadstate address and python version is found. Although it get rid of the error, pyflame is not collecting anything (only idle).

**Comment by max8899** [comment]
any update? same problem when tracing docker + nginx + uwsgi

**Comment by akatrevorjay** [comment]
It works against Py2 and Py3 engines running in uWSGI if you disable threads all together and just run a process pool, then pick a non-master worker process at random (at least in my cases on Precise, Xenial, and Yakkety using official python debs as well as the official python docker images). That said, my build does include a couple patches as it stands, but none come to mind that would fix this imo, I think the biggest thing here is just to make sure you are using only a pool of worker procs and not threads, hopefully that helps you guys at all!

**Comment by max8899** [comment]
Thanks for the reply, I'll try it out. And can I know why enable threads matters so much?

**Comment by eklitzke** [comment]
It should work with multi-threaded Python programs.

uWSGI is really weird because Python supports this embedded building mode, where you actually embed the Python interpreter into the executable you're building, and uWSGI can use this feature. When you build this way you can get some strange behavior because the compiler puts slightly different symbols into the executables and can do things like DCE to eliminate methods in surprising ways. So this is likely some corner case with how the version of uWSGI is built on the OS you're using.

I will try to get some time to reproduce this issue.

**Comment by zarnovican** [comment]
Hi @eklitzke, thanks for your time. I'm still able to reproduce it. Here is a complete list of steps from vanilla Ubuntu 16.04.1

```bash
sudo apt-add-repository ppa:trevorjay/pyflame
sudo apt-get update
sudo apt-get install pyflame
sudo apt-get install virtualenvwrapper build-essential python-dev -y 
. ~/.bashrc 
mkvirtualenv uwsgi
pip install flask uwsgi
cat >app.py <<EOF
from flask import Flask


app = Flask(__name__)


@app.route('/')
def hello_world():
    return 'Hello, World!'
EOF
```
```
uwsgi --workers 1 --module app:app --http-socket=127.0.0.1:8011 &

*** Starting uWSGI 2.0.14 (64bit) on [Sat Feb 18 13:04:18 2017] ***
compiled with version: 5.4.0 20160609 on 18 February 2017 13:03:37
os: Linux-4.4.0-59-generic #80-Ubuntu SMP Fri Jan 6 17:47:47 UTC 2017
...
*** Python threads support is disabled. You can enable it with --enable-threads ***
```
```
$ sudo pyflame `pgrep uwsgi`
Target is Python 0, which is not supported by this pyflame build.
```

**Comment by eklitzke** [comment]
It looks like GCC is being very aggressive about stripping unused symbols, and as a result is stripping `PyString_Type` from the uwsgi executable, which is how Pyflame can tell that the uwsgi binary is linked against Python 2.

There's a more correct way to tell the Python version using the method `Py_GetVersion`, which does exist in the uwsgi executable. However, doing remote calls from ptrace is sort of annoying and tricky, and something I've avoided so far. I will probably add a check for other `PyString_*` methods, but let me think a bit more about the best solution here.

**Comment by zarnovican** [comment]
If it complicates your code, you could just print an error when auto-detection fails. And add a pyflame command-line option for user to provide python version explicitly. People would understand..

**Comment by mmr** [comment]
There, implemented @zarnovican suggestion:
https://github.com/uber/pyflame/pull/54

It works.
Should buy some time while the detection is fixed

**Comment by eklitzke** [comment]
I merged #54 and plan to do a new release soon (once #55 is merged). In the mean time you can build from master.
